lettuce/io.py

```diff
@@ -61,3 +61,23 @@ def __call__(self, i, t, f):
                 print(err_u.item(), err_p.item(), file=self.out)
 
 
+class EnergyReporter:
+    """Reports the kinetic energy with respect to analytic solution."""
+    def __init__(self, lattice, flow, interval=1, out=sys.stdout):
+        self.lattice = lattice
+        self.flow = flow
+        self.interval = interval
+        self.out = [] if out is None else out
+        if not isinstance(self.out, list):
+            print("time      kinetic energy", file=self.out)
+
+    def __call__(self, i, t, f):
+        if t % self.interval == 0:
+            u = self.flow.units.convert_velocity_to_pu(self.lattice.u(f))
+            kinE= torch.sqrt(self.lattice.einsum("d,d->", [u,u])[self.lattice.field()])
+            nr_grid_points = self.lattice.rho(f).numel()
```

Comment:
the appropriate thing to do here (instead of dividing by the number of grid points) would be to include the physical length scale --

E = sum ( u_i u_i ) * dx**dim
with `dx = convert_length_to_pu(1.0)`

(Just realized that there is no `convert_length_to_pu` yet but that is trivial to add)

(or you could do the whole calculation in lettuce units and add energy conversion to the units module)

I would also add a method `energy` to the Lattice class - this is a calculation that you will reuse in other situations. Same for things like entropy, etc.  At some point we may port this into a separate module but for now the Lattice class should be the correct place.

Reply:
Ok, done.